### PR TITLE
Refactor network module - prepare it for easier sub-module testing

### PR DESCRIPTION
**Pull request by MehmedGIT**
Please feel free to suggest more changes that would lead to easier testing or improve the readability of the code. The Scrutinizer analysis already reports some good improvements so far.

Implemented refactoring changes:
- Makefile extension to support ocrd network test commands: `network-module-test`, `network-integration-test`
- Harmonize code - spacing between method parameters to maintain readability while still reducing the lines of the files. More variable typing wherever it was missing. Extensive use of `"` instead of `'` for strings to remove the need of escaping the `'` symbol when the string includes `'`.
- Improved error messages to include extra variable data if available for easier debugging
- Replaced repetitive constants with extensive use of Enums such as: `AgentType`, `DeployType`, `JobState`, `NetworkLoggingDirs`, `ServerApiTags`
- Env variable refactor to be generally used for all RabbitMQ clients and not just workers `OCRD_NETWORK_WORKER_QUEUE_CONNECT_ATTEMPTS` -> `OCRD_NETWORK_RABBITMQ_CLIENT_CONNECT_ATTEMPTS`
- The default job state is now set to `UNSET` when not provided
- `StateEnum` -> `JobState` for better clarification
- All occurrences of `status` -> `state` to maintain the same variable name everywhere
- The logging module to avoid potential name collisions: `src/ocrd_network/logging.py` -> `src/ocrd_network/logging_utils.py` 
- Combined all RabbitMQ-related data into a single dictionary variable instead of 5 separate variables (processing server and processing worker). The validators also return a dictionary of validated variables.
- Deployer module - increased readability and testability of separate deployments.
- Run-time data modules - created separate files for different concepts. Check [here](https://github.com/OCR-D/core/pull/1191/files#diff-fa00f886e24e9c1ba368197aeadfbbf87df96126f1db906a8aebec3e201250dc).
- Provided config parser wrapper methods for the different run-time data (for easier testing). Check [here](https://github.com/OCR-D/core/pull/1191/files#diff-5590f0a2244e9770feb1117d97a5dc0c15c5d9fff1a8a9cd53d927b30841e540).
- Processing server improvements (~20% less lines): 
  + Introduced `raise_http_exception` method to wrap repetitive lines
  + Created wrapper methods for separate concepts in big methods for easier concept testing, e.g., `parse_workflow_tasks`, `validate_first_task_input_file_groups_existence` and moved these to `utils` or `server_utils` depending on whether a server response is involved or not. Moreover, moved some other full methods which were better fitting inside `utils` or `server_utils` such as `create_processing_message`
  + Overall improved naming conventions for methods, e.g., `push_processor_job` -> `validate_and_forward_job_to_network_agent` where network agent is of kind `AgentType` (check Enums above)
  + Refactored server cache implementation + provided `sync` versions of the `async` methods to directly reuse these in the tests.
- Refactored rabbitmq_utils:
  + Moved the `create_queue` method from processing worker to `rabbitmq_utils` module
  + Provided extra 4 wrapper methods to remove code duplication in the processing server and processing worker. Check [here](https://github.com/OCR-D/core/pull/1191/files#diff-15f6a5de2f47765d32e3421c448829110d0aad0d900ae98a6635ef54cbd877e0).
- Extended the OcrdProcessingMessage schema to support `agent_type`. Check [here](https://github.com/OCR-D/core/pull/1191/files#diff-4d2a6af8df6f1a9df1841d5bfc120842a554c0eca2c600a09ea916ba1d180c6a).
- Fixed the OcrdResultMessage schema - the schema required either `path_to_mets` or `workspace_id`. However, setting `workspace_id` to `None` was triggering an error that it's not of type `string` according to the schema. On the other hand, assigning an empty string to `workspace_id` was triggering an error that either `path_to_mets` or `workspace_id` could be set but not both. So there was no way to create a result message without a crash. By default, the fields `path_to_mets` or `workspace_id` are set to empty strings when nothing is assigned. Check [here](https://github.com/OCR-D/core/pull/1191/files#diff-4436078efe91bb89b1dffcf849290a6e4e27f503e7ca47254103faab8ea43063).

Implemented tests:
- The naming conventions are `test_integration_*` or `test_modules_*` - the integration tests require access to either the MongoDB or RabbitMQ, but the module tests do not. Hence, the module tests can be executed without starting any extra services (faster for testing).
- Refactor `test_db.py` -> `test_integration_1_db.py` and replaced magic values with constants
- Refactor `test_rabbitmq.py` -> `test_integration_2_rabbitmq.py` and replaced magic values with constants
- `test_integration_3_server_cache_requests.py` - tests of the server cache for processing requests
- `test_integration_4_processing_worker.py` - tests of the processing worker 
- `test_integration_5_processing_server.py` - tests of the processing server
- `test_modules_logging_utils.py`
- `test_modules_param_validators.py`
- `test_modules_process_helpers.py` - Testing invocation of pythonic and bashlib processor
- `test_modules_server_cache_pages.py` - Unlike the server cache requests, the pages cache does not require database access.

Open issues:
- The bashlib processor test is still not working and is disabled currently - `_test_invoke_processor_bash()` - Shall be fixed in this PR.
- There is still the `deploy_agent_native_get_pid_hack` method, but now isolated away from the deployer. An eye should be kept on that in another PR.

Note: I have not merged the master branch to this PR yet, since v2.63.3 broke some mets caching. I will potentially do that before this PR can be finally merged.

Note2: The macOS tests are disabled since they fail due to a reason not related to this PR itself.

### CODE REVIEW

**Comment by MehmedGIT** [comment]
Arh, I hate Github Actions' false positives such as fb38c56 tests!

**Comment by MehmedGIT** [comment]
The workflow endpoint works just fine, however, there is a weird bug with the processing endpoint. The Pika library complains that the `queue_name` aka the `processor_name` is not a string but a `None`. It may be something coming from the updated RabbitMQ version - 3.12.

Request:
```shell
curl -X 'POST' \
  'http://localhost:8080/processor/run/ocrd-cis-ocropy-binarize' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "path_to_mets": "/home/mm/repos/ocrd_network_tests/ws29/data/mets.xml",
  "description": "The description of this execution",
  "input_file_grps": [
    "DEFAULT"
  ],
  "output_file_grps": [
    "OCR-D-BIN"
  ],
  "agent_type": "worker",
  "page_id": "PHYS_0001..PHYS_0003",
  "parameters": {}
}'
```

Response:
```python
INFO:     127.0.0.1:49952 - "POST /processor/run/ocrd-cis-ocropy-binarize HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/home/mm/venv38-all/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py", line 404, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/home/mm/venv38-all/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 84, in __call__
    return await self.app(scope, receive, send)
  File "/home/mm/venv38-all/lib/python3.8/site-packages/fastapi/applications.py", line 292, in __call__
    await super().__call__(scope, receive, send)
  File "/home/mm/venv38-all/lib/python3.8/site-packages/starlette/applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/mm/venv38-all/lib/python3.8/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/home/mm/venv38-all/lib/python3.8/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/home/mm/venv38-all/lib/python3.8/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
    raise exc
  File "/home/mm/venv38-all/lib/python3.8/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "/home/mm/venv38-all/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py", line 20, in __call__
    raise e
  File "/home/mm/venv38-all/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py", line 17, in __call__
    await self.app(scope, receive, send)
  File "/home/mm/venv38-all/lib/python3.8/site-packages/starlette/routing.py", line 718, in __call__
    await route.handle(scope, receive, send)
  File "/home/mm/venv38-all/lib/python3.8/site-packages/starlette/routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "/home/mm/venv38-all/lib/python3.8/site-packages/starlette/routing.py", line 66, in app
    response = await func(request)
  File "/home/mm/venv38-all/lib/python3.8/site-packages/fastapi/routing.py", line 273, in app
    raw_response = await run_endpoint_function(
  File "/home/mm/venv38-all/lib/python3.8/site-packages/fastapi/routing.py", line 190, in run_endpoint_function
    return await dependant.call(**values)
  File "/home/mm/repos/core/build/__editable__.ocrd-2.63.0-py3-none-any/ocrd_network/processing_server.py", line 461, in push_processor_job
    self.validate_agent_type_and_existence(processor_name=data.processor_name, agent_type=data.agent_type)
  File "/home/mm/repos/core/build/__editable__.ocrd-2.63.0-py3-none-any/ocrd_network/processing_server.py", line 451, in validate_agent_type_and_existence
    if self.network_agent_exists_worker(processor_name=processor_name):
  File "/home/mm/repos/core/build/__editable__.ocrd-2.63.0-py3-none-any/ocrd_network/processing_server.py", line 443, in network_agent_exists_worker
    return bool(self.check_if_queue_exists(processor_name=processor_name))
  File "/home/mm/repos/core/build/__editable__.ocrd-2.63.0-py3-none-any/ocrd_network/processing_server.py", line 379, in check_if_queue_exists
    self.rmq_publisher.create_queue(processor_name, passive=True)
  File "/home/mm/repos/core/build/__editable__.ocrd-2.63.0-py3-none-any/ocrd_network/rabbitmq_utils/publisher.py", line 57, in create_queue
    RMQConnector.queue_declare(
  File "/home/mm/repos/core/build/__editable__.ocrd-2.63.0-py3-none-any/ocrd_network/rabbitmq_utils/connector.py", line 187, in queue_declare
    queue = channel.queue_declare(
  File "/home/mm/venv38-all/lib/python3.8/site-packages/pika/adapters/blocking_connection.py", line 2512, in queue_declare
    validators.require_string(queue, 'queue')
  File "/home/mm/venv38-all/lib/python3.8/site-packages/pika/validators.py", line 15, in require_string
    raise TypeError('%s must be a str or unicode str, but got %r' % (
TypeError: queue must be a str or unicode str, but got None
```

File: src/ocrd_network/constants.py [hunk 1]
```diff
@@ -1,2 +1,55 @@
-NETWORK_AGENT_SERVER = 'server'
-NETWORK_AGENT_WORKER = 'worker'
+from enum import Enum
+
+DOCKER_IMAGE_MONGO_DB = "mongo"
+DOCKER_IMAGE_RABBIT_MQ = "rabbitmq:3.12-management"
+# These feature flags are required by default to use the newer version
+DOCKER_RABBIT_MQ_FEATURES = "quorum_queue,implicit_default_bindings,classic_mirrored_queue_version"
+
+NETWORK_API_TAG_ADMIN = "admin"
```
Comment (kba) [inline src/ocrd_network/constants.py:8]: make these enums as well, while you're at it?

File: src/ocrd_network/runtime_data/network_agents.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+from ..constants import AgentType, DeployType
+
+
+# TODO: Find appropriate replacement for the hack
+def deploy_agent_native_get_pid_hack(logger: Logger, ssh_client, start_cmd: str):
+    channel = ssh_client.invoke_shell()
+    stdin, stdout = channel.makefile("wb"), channel.makefile("rb")
+    logger.debug(f"Executing command: {start_cmd}")
+
+    # TODO: This hack should still be fixed
```
Comment (kba) [inline src/ocrd_network/runtime_data/network_agents.py:14]: Do we have an issue tracking this?

File: src/ocrd_network/server_utils.py [hunk 3]
```diff
@@ -1,94 +1,168 @@
...
     return FileResponse(path=log_file_path, filename=log_file_path.name)
 
 
-async def validate_and_return_mets_path(logger, job_input: PYJobInput) -> str:
-    # This check is done to return early in case the workspace_id is provided
-    # but the abs mets path cannot be queried from the DB
-    if not job_input.path_to_mets and job_input.workspace_id:
+async def get_workflow_content(logger: Logger, workflow_id: str, workflow: Union[UploadFile, None]) -> str:
+    if not workflow and not workflow_id:
+        message = "Either 'workflow' binary or 'workflow_id' must be provided. Both are missing."
```
Comment (kba) [inline src/ocrd_network/server_utils.py:94]:
```suggestion
        message = "Either 'workflow' must be uploaded as a fie or 'workflow_id' must be provided. Both are missing."
```

File: src/ocrd_network/utils.py [hunk 4]
```diff
@@ -36,6 +39,24 @@ def calculate_execution_time(start: datetime, end: datetime) -> int:
     return int((end - start).total_seconds() * 1000)
 
 
+def convert_url_to_uds_format(url: str) -> str:
+    return f"http+unix://{url.replace('/', '%2F')}"
+
+
+def expand_page_ids(page_id: str) -> List:
```
Comment (kba) [inline src/ocrd_network/utils.py:46]: Move this to `ocrd_utils` and use it in `OcrdMets.get_physical_pages`, too?

**Review by kba — APPROVED** [review]
Just a quick browse through, since this is quite extensive. Much improved I think!

File: src/ocrd_network/logging.py [hunk 5]
```diff
@@ -1,15 +1,7 @@
 from pathlib import Path
```
[the remark this answers is not in the document]
Comment (MehmedGIT) [reply src/ocrd_network/logging.py:1]: Ah, nice hint. I have already eliminated some shadowed variables. Will pay attention to files as well!

File: src/ocrd_network/constants.py (hunk 1 above)
Comment (MehmedGIT) [reply src/ocrd_network/constants.py:8]: Network API tags, sure. Still not sure about docker images and features though.

File: src/ocrd_network/runtime_data/network_agents.py (hunk 2 above)
Comment (MehmedGIT) [reply src/ocrd_network/runtime_data/network_agents.py:14]: Not sure - maybe not. I am planning to rework this after the module tests where it will be easier for me to try different things and just run module tests to verify different solutions.

File: src/ocrd_network/utils.py (hunk 4 above)
Comment (MehmedGIT) [reply src/ocrd_network/utils.py:46]: Sure, I could move it to `ocrd_utils`. However, I am not sure whether it is a good idea to move it inside `OcrdMets.get_physical_pages`. That method seems very complex. I would suggest splitting it few simpler methods and use `ge_physical_pages` as a wrapper over the simpler methods.

File: src/ocrd_network/processor_server.py [hunk 6]
```diff
@@ -76,74 +66,70 @@ def __init__(self, mongodb_addr: str, processor_name: str = "", processor_class=
...
+    async def on_shutdown(self) -> None:
+        """
+        TODO: Perform graceful shutdown operations here
+        """
+        pass
 
-        # Create routes
-        self.router.add_api_route(
-            path='/info',
+    def add_api_routes_processing(self):
```
Comment (joschrew) [inline src/ocrd_network/processor_server.py:80]: I think a call of this method probably in the constructor is missing. Or anywhere else, it is not used anywhere.

**Review by joschrew — APPROVED** [review]
Very big pull request, you obviously did a lot of work here, respect for that. For me this PR is hard to review, because this are really a lot of changes. I have tried to look into all the files but especially the changes in the processing_server where hard to follow. I am sure I could easily have missed some stuff.
I run the tests though and I tried to run the changes in a self build Dockercontainer together with a pythonic and bashlib worker. This tests all passed on my machine.
The good thing is this changes nearly only affect the network package, so I am fine with merging them, even if no very detailed review can be done here.

**Comment by MehmedGIT** [comment]
> For me this PR is hard to review, because this are really a lot of changes.

Understandable. 

> I have tried to look into all the files but especially the changes in the processing_server where hard to follow. 

For the processing server specifically, it is better to check the result not the changes. That one file got a lot of modifications and had to become fewer lines (~20% less).

> I run the tests though and I tried to run the changes in a self build Dockercontainer together with a pythonic and bashlib worker. This tests all passed on my machine.

Have you also tried the `_test_invoke_processor_bash()` test by removing the underscore? That is the one I am not sure how to run yet to test the spawning of a bashlib processor separately from a worker.

> The good thing is this changes nearly only affect the network package

Yes, tried to avoid moving stuff inside other modules (shall be the last thing to do) for now since this PR is already big enough.

File: src/ocrd_network/processor_server.py (hunk 6 above)
Comment (MehmedGIT) [reply src/ocrd_network/processor_server.py:82]: Good catch. There are not any processor server tests yet so I missed that.

File: src/ocrd_network/runtime_data/hosts.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+        # Key: processor_name, Value: list of ports
+        self.processor_servers_ports: dict = {}
+
+    def __add_deployed_agent_server_port_to_cache(self, processor_name: str, port: int) -> None:
+        if processor_name not in self.processor_servers_ports:
+            self.processor_servers_ports[processor_name] = [port]
+            return
+        self.processor_servers_ports[processor_name] = self.processor_servers_ports[processor_name].append(port)
+
+    def __append_network_agent_to_lists(self, agent_data: DataNetworkAgent) -> None:
```
Comment (kba) [inline src/ocrd_network/runtime_data/hosts.py:54]: Why the double underscores? If something starts with dunder `__`, I expect it to be some sort of official internal.

**Review by kba — APPROVED** [review]
LGTM, very thorough. I'll now try to fix the bashlib test case, besides that looks good to go.

File: src/ocrd_network/runtime_data/hosts.py (hunk 7 above)
Comment (MehmedGIT) [reply src/ocrd_network/runtime_data/hosts.py:54]: To hint that it is an internal helper method and must not be called outside the class.

Comment (joschrew) [reply src/ocrd_network/runtime_data/hosts.py:54]: I thought about the single and double leading underscores of function declarations too. Maybe I also was somebody who introduced them in some places. Previously I thought it was to indicate kind of private functions, but I googled a bit and came to the conclusion there is no one single way how this is understood. Tbh I think I would remove all these leading underscores (single and double) from "normal" functions and methods but I don't care too much.

Comment (kba) [reply src/ocrd_network/runtime_data/hosts.py:54]: Not that important, as long it's not a "real" dunder method (ie. __foo__, with double underscores leading and trailing).
